runtime/lua/lsp/structures.lua

```diff
@@ -0,0 +1,191 @@
+-- luacheck: globals vim
+
+local lsp_util = require('lsp.util')
+
+local protocol = require('lsp.protocol')
+
+-- Helper functions
+local check_table = function (t)
+  if type(t) ~= 'table' then
+    t = {}
+  end
+
+  return t
+end
+
+-- Structure definitions
+local structures = {}
+
+structures.EOL = function()
+  if vim.api.nvim_buf_get_option(0, 'eol') then
+    return "\n"
+  else
+    return ''
+  end
+end
+
+structures.DocumentUri = function(args)
+  return args
+    or 'file://' .. vim.api.nvim_buf_get_name(0)
+end
+structures.languageId = function(args)
+  return args
+     or vim.api.nvim_buf_get_option(0, 'filetype')
+end
+-- TODO: Increment somehow
+local __version = 0
+structures.version = function(args)
+  __version = __version + 1
+  return args
+    or __version
+end
+structures.text = function(args)
+  return args
+    or lsp_util.get_buffer_text(0)
+end
+structures.TextDocumentIdentifier = function(args)
+  args = check_table(args)
+
+  return {
+    uri = structures.DocumentUri(args.uri),
+  }
+end
+structures.VersionedTextDocumentIdentifier = function(args)
+  args = check_table(args)
+
+  local result = structures.TextDocumentIdentifier(args)
+  result.version = structures.version(args.version)
+
+  return result
+end
+structures.TextDocumentItem = function(args)
+  args = check_table(args)
+
+  return {
+    uri = structures.DocumentUri(args.uri),
+    languageId = structures.languageId(args.languageId),
+    version = structures.version(args.version),
+    text = structures.text(args.text),
+  }
+end
+structures.line = function(args)
+  return args
+    -- TODO: Check the conversion of some of these functions from nvim <-> lua
+    or (vim.api.nvim_call_function('line', {'.'}) - 1)
+end
+structures.character = function(args)
+  return args
+    or (vim.api.nvim_call_function('col', {'.'}) - 1)
+end
+
+structures.Position = function(args)
+  args = check_table(args)
+
+  return {
+    line = structures.line(args.line),
+    character = structures.character(args.character),
+  }
+end
+structures.TextDocumentPositionParams = function(args)
+  args = check_table(args)
+
+  return {
+    textDocument = structures.TextDocumentIdentifier(args.textDocument),
+    position = structures.Position(args.position),
+  }
+end
+structures.ReferenceContext = function(args)
+  args = check_table(args)
+
+  return {
+    includeDeclaration = args.includeDeclaration or true,
+  }
+end
+structures.ReferenceParams = function(args)
+  args = check_table(args)
+
+  local positionParams = structures.TextDocumentPositionParams(args)
+  positionParams.context = structures.ReferenceContext(args.context)
+
+  return positionParams
+end
+structures.DidOpenTextDocumentParams = function(args)
```

Comment:
I do not think it makes sense to both have `structures` methods for concrete messages like `structures.DidOpenTextDocumentParams` that then is only used in one method like `notifications.textDocument.didOpen` (which also commonly are completely empty except for the "structures" call) . This is just indirection and multiple naming schemes for the same thing, not encapsulation. Probably `structures` should only be for the reuseable structure bits, and we merge these duplicate methods into their respective request/notification handlers.

Reply:
@bfredl 
I would like to remove the `request / notification` validation and use structures only as a request parameter constructor to avoid duplication.
Since validation has to keep up with changes in language server protocol, I think it is inconvenient for users to wait for neovim validation to be fixed before the release of neovim.
What do you think about this?

Reply:
I agree, the current approach is weird indirection. Instead of "validating" inside already made requests, we should just have wrapper functions around the (public) request API.

Reply:
> we should just have wrapper functions around the (public) request API.

How do you expect to wrap?

Reply:
I mean functions that provide a higher-level interface around the low-level API. Currently `vim.lsp.request("some/request", params)` will implicitly transform the arguments according to rules specific for each request, I understood you want to get rid of that (and I agree). Instead the same functionality can be expressed as high-level functions that would construct the JSON arguments and then call `vim.lsp.request`. This will be an addition to the low-level API instead of being entangled with it (and simpler design overall).

Completion might be a good example, instead of
```
    let params = luaeval("vim.lsp.request.completion("
                         \ . "{ position = { character = _A }})",
                         \  col('.') + len(a:base))
    let results = lsp#request('textDocument/completion', params)
```

we could use (not sure if the arguments are the right ones, I mean the general form)

```
   let result = luaeval('vim.lsp.xxx.completion(_A)', col('.') + len(a:base))`
```

where `vim.lsp.xxx.completion()` (not sure of the proper module path) would invoke `vim.lsp.request("textDocument/completion", {params})` with the correct params, perhaps transform the result, etc.

Reply:
I see. Thank you for your explanation. :)
I try it.

Reply:
@bfredl 
I've forgotten handling server capabilities.
In the original implementation, at the time of verifying the request parameter before requesting, for the function whose language server capability is false, only writing the log, but it was not actually requested.
But now, I remove verifying the request parameter steps.
I think we should leave only checking language server capabilities before requesting.